### PR TITLE
Fix/garbage allocation

### PR DESCRIPTION
**Pull request by crsib**
Prevents the garbage allocation in the cases the CommandBinder is not used

### CODE REVIEW

**Comment by WillzZz** [comment]
Cool, I'll have to take a better look at this when I get a minute. It makes sense that we alloc too much with the delegates. Have you profiled with a test project to verify we're making an improvement here?

**Comment by crsib** [comment]
It was proven by Unity Profiler that zero garbage is generated if only normal delegates are added to the signal (i.e. via AddListener). In other cases - yep, it makes sense that lots of allocations are performed

File: StrangeIoC/scripts/strange/extensions/signal/impl/BaseSignal.cs [hunk 1]
```diff
@@ -33,45 +34,60 @@ public class BaseSignal : IBaseSignal
...
 			OnceBaseListener += callback;
 		}
 
-		public void RemoveListener(Action<IBaseSignal, object[]> callback) { BaseListener -= callback; }
+		public void RemoveListener(Action<IBaseSignal, object[]> callback) 
+		{
+			var listeners = BaseListener;
+			BaseListener -= callback;
+
+			if (Object.ReferenceEquals(BaseListener, listeners))
```
Comment (WillzZz) [inline StrangeIoC/scripts/strange/extensions/signal/impl/BaseSignal.cs]: In what scenario is this true?

File: StrangeIoC/scripts/strange/extensions/signal/impl/BaseSignal.cs [hunk 2]
```diff
@@ -33,45 +34,60 @@ public class BaseSignal : IBaseSignal
 	{
         
 		/// The delegate for repeating listeners
-		public event Action<IBaseSignal, object[]> BaseListener = delegate { };
+		private event Action<IBaseSignal, object[]> BaseListener = delegate { };
+		private int BaseListenerSubsrcibersCount = 0;
```
Comment (WillzZz) [inline StrangeIoC/scripts/strange/extensions/signal/impl/BaseSignal.cs]: Typo

File: StrangeIoC/scripts/strange/extensions/signal/impl/Signal.cs [hunk 3]
```diff
@@ -121,40 +136,60 @@ public void AddListener(Action<T> callback)
...
+		}
     }
 
     /// Base concrete form for a Signal with two parameters
     public class Signal<T, U> : BaseSignal
     {
-        public event Action<T, U> Listener = delegate { };
-        public event Action<T, U> OnceListener = delegate { };
+        public event Action<T, U> Listener = delegate { };
+		public event Action<T, U> OnceListener;
```
Comment (WillzZz) [inline StrangeIoC/scripts/strange/extensions/signal/impl/Signal.cs]: Minor formatting nit

File: StrangeIoC/scripts/strange/extensions/signal/impl/Signal.cs [hunk 4]
```diff
@@ -177,26 +213,44 @@ public override List<Type> GetTypes()
...
+			if(base.NonEmpty)
+			{
+				object[] outv = { type1, type2 };
+				base.Dispatch(outv);
+			}
         }
         private Action<T, U> AddUnique(Action<T, U> listeners, Action<T, U> callback)
-        {
+        {
+			if (null == listeners)
```
Comment (WillzZz) [inline StrangeIoC/scripts/strange/extensions/signal/impl/Signal.cs]: Another formatting nit. Probably you're using spaces and we're using tabs.

File: StrangeIoC/scripts/strange/extensions/signal/impl/BaseSignal.cs [hunk 5]
```diff
@@ -33,45 +34,60 @@ public class BaseSignal : IBaseSignal
...
 
 		public void Dispatch(object[] args) 
 		{ 
 			BaseListener(this, args);
-			OnceBaseListener(this, args);
-			OnceBaseListener = delegate { };
+
+			if(null != OnceBaseListener)
+			{
+				var onceBaseDelegates = OnceBaseListener;
```
Comment (WillzZz) [inline StrangeIoC/scripts/strange/extensions/signal/impl/BaseSignal.cs]: This (and related) are to allow a method to addOnce itself during the call?

**Comment by WillzZz** [comment]
Definitely clean up some of the formatting, but in general it looks like a good addition. Thanks!

**Comment by fogAndWhisky** [comment]
We've added a TODO to review this PR and get it into the next version. Sorry for the delay.

**Comment by WillzZz** [comment]
Thanks for the push in the right direction! This is a good change. I took a look here and made my own fixes right along these lines. Closing.

**Pull request closed without merging**